### ISSUE TITLE
[Cpp] Runtime build failure on gcc 6.3.1

### ISSUE THREAD

**Issue by Pursuit92** [antlr/antlr4#1608]
After a gcc upgrade, I'm finding that it fails to compile the runtime with this error: 

```
In file included from /usr/include/c++/6.3.1/exception:171:0,
                 from /usr/include/c++/6.3.1/new:40,
                 from /usr/include/c++/6.3.1/bits/stl_construct.h:59,
                 from /usr/include/c++/6.3.1/bits/stl_tempbuf.h:60,
                 from /usr/include/c++/6.3.1/bits/stl_algo.h:62,
                 from /usr/include/c++/6.3.1/algorithm:62,
                 from .../antlr4-runtime/runtime/src/antlr4-common.h:8,
                 from .../antlr4-runtime/runtime/src/tree/pattern/ParseTreePattern.h:8,
                 from .../antlr4-runtime/runtime/src/tree/pattern/ParseTreePatternMatcher.cpp:6:
/usr/include/c++/6.3.1/bits/nested_exception.h: In instantiation of ‘void std::throw_with_nested(_Tp&&) [with _Tp = const char (&)[25]]’:
.../antlr4-runtime/runtime/src/tree/pattern/ParseTreePatternMatcher.cpp:103:54:   required from here
/usr/include/c++/6.3.1/bits/nested_exception.h:121:7: error: static assertion failed: throw_with_nested argument must be CopyConstructible
       static_assert(_CopyConstructible::value,
       ^~~~~~~~~~~~~
make[2]: *** [runtime/CMakeFiles/antlr4_shared.dir/build.make:2895: runtime/CMakeFiles/antlr4_shared.dir/src/tree/pattern/ParseTreePatternMatcher.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:90: runtime/CMakeFiles/antlr4_shared.dir/all] Error 2
make: *** [Makefile:128: all] Error 2
```

The fedora 23/24 docker containers can be used to test this - after installing the "Development Tools" group, gcc-c++, cmake, and libuuid-devel, fedora 23 builds the runtime just fine while 24 fails with the above error.

**Comment by jwakely** [comment]
I added the check for CopyConstructible to GCC's `throw_with_nested` because that's what the standard requires, but it looks like there could be a defect in the standard, because it forbids this case.

You can workaround it by throwing a pointer instead of a string literal:

    std::throw_with_nested((const char*)"Cannot invoke start rule");

**Comment by jwakely** [comment]
I've created https://gcc.gnu.org/PR79114 to deal with this.

**Comment by gaulouis** [comment]
jwakely :+1:

**Comment by mike-lischke** [comment]
@jwakely I believe this issue can be closed then.

**Comment by jwakely** [comment]
You could add the cast if you want to ensure antlr compiles with all versions of GCC. The fixed GCC has not been released yet, and the 6.3 release will always have the problem.
